packages/core/src/command/rate-limiting/ddb-store.ts

```diff
@@ -0,0 +1,323 @@
+   *
+   * @public
+   */
+  async decrement(key: string): Promise<void> {
+    const { log } = out(`DynamoStore decrement - key ${key}`);
+    const item = await this.ddbUtils.update({
+      partition: this.prefixCountKey(key),
+      expression: "ADD totalHits :inc",
+      expressionAttributesValues: {
+        ":inc": -1,
+      },
+    });
+    const entry = rateLimitCountSchema.safeParse(item.Attributes);
+    if (!entry.success) {
+      const error = entry.error;
+      const msg = "Error parsing DynamoStore count entry";
+      log(`${msg} - error: ${errorToString(error)}`);
+      capture.error(msg, {
+        extra: {
+          key,
+          context: "dynamo-store.decrement",
+          error,
+        },
+      });
+      throw new MetriportError(msg, error, { method: "decrement", key });
+    }
+  }
+
+  /**
+   * Method to reset a client's hit counter.
+   *
+   * @param key {string} - The identifier for a client.
+   *
+   * @public
+   */
+  async resetKey(key: string): Promise<void> {
+    const { log } = out(`DynamoStore resetKey - key ${key}`);
+    const item = await this.ddbUtils.update({
+      partition: this.prefixCountKey(key),
+      expression: "SET totalHits = :totalHits REMOVE resetTime",
+      expressionAttributesValues: {
+        ":totalHits": 0,
+      },
+    });
+    const entry = rateLimitCountSchema.safeParse(item.Attributes);
+    if (!entry.success) {
+      const error = entry.error;
+      const msg = "Error parsing DynamoStore count entry";
+      log(`${msg} - error: ${errorToString(error)}`);
+      capture.error(msg, {
+        extra: {
+          key,
+          context: "dynamo-store.resetKey",
+          error,
+        },
+      });
+      throw new MetriportError(msg, undefined, { method: "resetKey", key });
+    }
+  }
+
+  /**
+   * Method to fetch a client's limit. Introduced by Metriport.
+   *
+   * @param key {string} - The identifier for a client.
+   *
+   * @returns {number} - The limit of hits within the window.
+   *
+   * @public
+   */
+  async getLimit(key: string, defaultLimit: number): Promise<number> {
+    const { log } = out(`DynamoStore get - key ${key}`);
+    const item = await this.getOrCreateLimit(key, defaultLimit);
+    const entry = rateLimitLimitSchema.safeParse(item);
+    if (!entry.success) {
+      const error = entry.error;
+      const msg = "Error parsing DynamoStore limit entry";
+      log(`${msg} - error: ${errorToString(error)}`);
+      capture.error(msg, {
+        extra: {
+          key,
+          context: "dynamo-store.getLimit",
+          error,
+        },
+      });
+      throw new MetriportError(msg, error, { method: "getLimit", key, defaultLimit });
+    }
+    return entry.data.windowLimit;
+  }
+
+  async getOrCreateLimit(key: string, defaultLimit: number): Promise<DocumentClient.AttributeMap> {
+    const item = await this.ddbUtils.get({ partition: this.prefixLimitKey(key) });
+    if (!item.Item) {
```

Comment:
nit: NABD really, but I find it easier to avoid multiple nested scopes - could decrease one nesting if we do like so:

```ts
if (item.Item) return item.Item;
// logic when the item is not set
// ...
```

Reply:
good call!

Reply:
done.